devdoc/channel_2.md

```diff
@@ -0,0 +1,72 @@
+# channel_2 requirements:
+
+## Overview
+
+The current implementation of `channel` has certain pitfalls. `channel_2` aims to address these pitfalls by making the following changes:
```

Comment:
>channel_2` aims to address these pitfalls by making the following changes: [](http://example.com/codeflow?start=63&length=74)

Why did you decide to make a new API instead of fixing the existing one?  This doc makes it sound like we will have the original `channel.h` API and a new `channel_2.h` API and they will live side-by-side.  Will the old one ever be removed?  When?

This also makes me wonder how the existing code, for `channel_pull` and `channel_push` is related to the new code for `channel_2_pull` and `channel_2_push`? Are you going to completely scrap the existing implementation? Are you going to have some sort of shim later so `channel_2_push` internally calls `channel_push` or vise versa?

If `channel_2.h` is a complete rewrite, does that also mean a new set of requirements and tests for all of the existing functionality?

From what I see here, it looks like you're writing a completely new channel implementation instead of fixing the old one. Is this right? I hope not.

I would suggest making this into a set of incremental improvements to the existing channel, where each one is an isolated piece of work that can completed and verified before the next one is started. For example, you could remove `async_op` as step 1, finish it, propagate it up to geo and make the changes there, then go back and remove channel_internal, propagate it up, and so on.